### PR TITLE
Add cmake build support (w/lots of diagnostics enabled)

### PR DESCRIPTION
**Pull request by larseggert**

### CODE REVIEW

**Comment by jariarkko** [comment]
Thanks again for this, Lars. We will be looking at this in detail and doing the change (but both of us have a bit of a learning curve with cmake, so bear with us).

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+if(IWYU)
+  message(STATUS "Found include-what-you-use: ${IWYU}")
+  set(CMAKE_C_INCLUDE_WHAT_YOU_USE ${IWYU})
+endif()
+
+find_program(CPPCHECK NAMES cppcheck)
+if(CPPCHECK)
+  message(STATUS "Found cppcheck: ${CPPCHECK}")
+  set(CMAKE_C_CPPCHECK ${CPPCHECK} -q --inline-suppr --enable=all)
+  set(CMAKE_CXX_CPPCHECK ${CMAKE_C_CPPCHECK})
```
Comment (larseggert) [inline CMakeLists.txt]: This line should be removed (since you don't use C++)

**Comment by jariarkko** [comment]
We have not forgotten this, but I've had to implement some other things before this can be done in its current form. I've rewritten some parts of the code to make it portable across platforms and at the same time remove the padding and alignment related warnings. More work on cleaning up the rest of the warnings is still in progress.

**Comment by jariarkko** [comment]
There are, however, some issues that we will need to resolve in this specific PR as well.

1/ For some reason the code to check what compiler options are usable doesn't work. I think this is an Apple OS X thing; their gcc is older and has slightly different syntax. I could update, but I want the software to work in everyone's environment. One way would be to remove the options that lead to this problem.

2/ Some warning options will lead to problematic results. Specifically, -Wcovered-switch-default will prevent defensive coding style, as default branches will be prohibited for enum values. Perhaps this is a philosophical issue but I like my code to detect problems and fail visibly if something's wrong, even if the reason for such a problem would have to be memory corruption or pointer error. I plan to disable the warning and use -Wno-covered-switch-default. There may be other warnings, I have not gotten through all of them yet.

3/ I would like to retain the ability to run tests as part of the regular make run; it has been very useful in catching errors. How do I do that? I still need to learn more about cmake...

**Comment by larseggert** [comment]
I'm surprised the compiler option checker doesn't work on OSX with Apple's gcc, let me investigate.

Yes, please add whatever `-Wno-*` options make sense for you - I do the same for my projects.

Regarding testing, it's very easy to add test targets, see https://gitlab.kitware.com/cmake/community/wikis/doc/ctest/Testing-With-CTest

**Comment by larseggert** [comment]
The compiler option stuff does work for me on OSX

**Comment by jariarkko** [comment]
Thanks. Will look at testing; warnings should be pretty much handled by now.

regarding the compiler options, what i get is:

Jaris-MacBook-Pro:spindump jariarkko$ git clone https://github.com/larseggert/spindump.git
Jaris-MacBook-Pro:spindump jariarkko$ cd spindump
Jaris-MacBook-Pro:spindump jariarkko$ cmake .
-- The C compiler identification is AppleClang 9.0.0.9000039
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Performing Test _pipe
-- Performing Test _pipe - Success
-- Performing Test _Wextra
-- Performing Test _Wextra - Success
-- Performing Test _Wpedantic
-- Performing Test _Wpedantic - Success
-- Performing Test _Weverything
-- Performing Test _Weverything - Success
-- Performing Test _Werror
-- Performing Test _Werror - Success
-- Performing Test _fcolor_diagnostics
-- Performing Test _fcolor_diagnostics - Success
-- Performing Test _fdiagnostics_color_always
-- Performing Test _fdiagnostics_color_always - Success
-- Performing Test _fsanitize_memory
-- Performing Test _fsanitize_memory - Failed
-- Performing Test _fsanitize_memory_track_origins
-- Performing Test _fsanitize_memory_track_origins - Success
-- Performing Test _fsanitize_memory_use_after_dtor
-- Performing Test _fsanitize_memory_use_after_dtor - Success
-- Performing Test _fsanitize_address
-- Performing Test _fsanitize_address - Failed
-- Performing Test _fsanitize_undefined
-- Performing Test _fsanitize_undefined - Success
-- Performing Test _fno_omit_frame_pointer
-- Performing Test _fno_omit_frame_pointer - Success
-- Performing Test _fsanitize_address_use_after_scope
-- Performing Test _fsanitize_address_use_after_scope - Success
-- Performing Test _fno_optimize_sibling_calls
-- Performing Test _fno_optimize_sibling_calls - Success
-- Performing Test _fno_common
-- Performing Test _fno_common - Success
-- Performing Test _fsanitize_unsigned_integer_overflow
-- Performing Test _fsanitize_unsigned_integer_overflow - Success
-- Performing Test _fsanitize_implicit_conversion
-- Performing Test _fsanitize_implicit_conversion - Failed
-- Performing Test _fsanitize_nullability
-- Performing Test _fsanitize_nullability - Success
-- Found PkgConfig: /opt/local/bin/pkg-config (found version "0.29.2") 
-- Found Curses: /usr/lib/libcurses.dylib  
-- Checking for module 'libpcap'
--   Found libpcap, version 1.9.0-PRE-GIT
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/jariarkko/LME/Spindump/spindump.cmake2/spindump
Jaris-MacBook-Pro:spindump jariarkko$ make
Scanning dependencies of target libspindump
[  2%] Building C object src/CMakeFiles/libspindump.dir/spindump_analyze.c.o
clang: error: argument unused during compilation: '-fsanitize-memory-track-origins' [-Werror,-Wunused-command-line-argument]
clang: error: argument unused during compilation: '-fsanitize-memory-use-after-dtor' [-Werror,-Wunused-command-line-argument]
clang: error: argument unused during compilation: '-fsanitize-address-use-after-scope' [-Werror,-Wunused-command-line-argument]
make[2]: *** [src/CMakeFiles/libspindump.dir/spindump_analyze.c.o] Error 1
make[1]: *** [src/CMakeFiles/libspindump.dir/all] Error 2
make: *** [all] Error 2

**Comment by larseggert** [comment]
I get these `-Wunused-command-line-argument` errors, too, but *only* when there are other errors in a file. (I think it's a bug in clang.) For `spindump_analyze.c`, here are all the errors I get:
```
Scanning dependencies of target libspindump
[  2%] Building C object src/CMakeFiles/libspindump.dir/spindump_analyze.c.o
Warning: include-what-you-use reported diagnostics:
error: unsupported argument 'nullability' to option 'fsanitize='
error: invalid argument '-fsanitize-address-use-after-scope' only allowed with '-fsanitize=address'
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:352:3: error: default label in switch which covers all enumeration values [-Werror,-Wcovered-switch-default]
  default:
  ^

/Users/lars/Documents/Code/spindump/src/spindump_analyze.h should add these lines:
#include <_types/_uint16_t.h>  // for uint16_t
#include <_types/_uint8_t.h>   // for uint8_t
#include "spindump_util.h"     // for spindump_address
struct spindump_connection;
struct spindump_packet;

/Users/lars/Documents/Code/spindump/src/spindump_analyze.h should remove these lines:
- #include "spindump_connections_structs.h"  // lines 27-27
- #include "spindump_table.h"  // lines 28-28

The full include-list for /Users/lars/Documents/Code/spindump/src/spindump_analyze.h:
#include <_types/_uint16_t.h>  // for uint16_t
#include <_types/_uint8_t.h>   // for uint8_t
#include "spindump_capture.h"  // for spindump_capture_linktype
#include "spindump_util.h"     // for spindump_address
struct spindump_analyze;  // lines 59-59
struct spindump_connection;
struct spindump_packet;
---

/Users/lars/Documents/Code/spindump/src/spindump_analyze.c should add these lines:
#include <_types/_uint32_t.h>                   // for uint32_t
#include <netinet/in.h>                         // for IPPROTO_ICMP, IPPROTO...
#include <stdlib.h>                             // for free, malloc
#include <sys/socket.h>                         // for AF_INET, AF_INET6
#include "spindump_connections_structs.h"       // for spindump_connection
#include "spindump_packet.h"                    // for spindump_packet
#include "spindump_protocols.h"                 // for spindump_ip, spindump...
#include "spindump_stats.h"                     // for spindump_stats, spind...
#include "spindump_table.h"                     // for spindump_connectionst...
#include "spindump_table_structs.h"             // for spindump_connectionst...

/Users/lars/Documents/Code/spindump/src/spindump_analyze.c should remove these lines:
- #include <netinet/icmp6.h>  // lines 26-26
- #include <netinet/ip.h>  // lines 24-24
- #include <netinet/ip_icmp.h>  // lines 25-25
- #include "spindump_analyze_quic.h"  // lines 34-34
- #include "spindump_connections_set.h"  // lines 29-29

The full include-list for /Users/lars/Documents/Code/spindump/src/spindump_analyze.c:
#include "spindump_analyze.h"
#include <_types/_uint32_t.h>                   // for uint32_t
#include <netinet/in.h>                         // for IPPROTO_ICMP, IPPROTO...
#include <stdlib.h>                             // for free, malloc
#include <string.h>                             // for memset, memcpy
#include <sys/socket.h>                         // for AF_INET, AF_INET6
#include "spindump_analyze_aggregate.h"         // for spindump_analyze_proc...
#include "spindump_analyze_icmp.h"              // for spindump_analyze_proc...
#include "spindump_analyze_tcp.h"               // for spindump_analyze_proc...
#include "spindump_analyze_udp.h"               // for spindump_analyze_proc...
#include "spindump_connections.h"               // for spindump_connections_...
#include "spindump_connections_set_iterator.h"  // for spindump_connection_s...
#include "spindump_connections_structs.h"       // for spindump_connection
#include "spindump_packet.h"                    // for spindump_packet
#include "spindump_protocols.h"                 // for spindump_ip, spindump...
#include "spindump_stats.h"                     // for spindump_stats, spind...
#include "spindump_table.h"                     // for spindump_connectionst...
#include "spindump_table_structs.h"             // for spindump_connectionst...
#include "spindump_util.h"                      // for spindump_assert, spin...
---

warning: argument unused during compilation: '-fsanitize-address-use-after-scope' [clang-diagnostic-unused-command-line-argument]
warning: argument unused during compilation: '-fsanitize-memory-track-origins' [clang-diagnostic-unused-command-line-argument]
warning: argument unused during compilation: '-fsanitize-memory-use-after-dtor' [clang-diagnostic-unused-command-line-argument]
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:23:1: warning: #includes are not sorted properly [llvm-include-order]
#include <string.h>
^        ~~~~~~~~~~
         "spindump_analyze.h"
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:352:3: warning: default label in switch which covers all enumeration values [clang-diagnostic-covered-switch-default]
  default:
  ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:522:31: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
  unsigned int ipHeaderSize = SPINDUMP_IP_HL(&ip)*4;
                              ^
/Users/lars/Documents/Code/spindump/src/spindump_protocols.h:62:32: note: expanded from macro 'SPINDUMP_IP_HL'
#define SPINDUMP_IP_HL(ip)    (((ip)->ip_vhl) & 0x0f)
                               ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:530:23: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
  uint8_t ipVersion = SPINDUMP_IP_V(&ip);
                      ^
/Users/lars/Documents/Code/spindump/src/spindump_protocols.h:63:32: note: expanded from macro 'SPINDUMP_IP_V'
#define SPINDUMP_IP_V(ip)     (((ip)->ip_vhl) >> 4)
                               ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:533:80: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    spindump_warnf("IP versions inconsistent in Ethernet frame and IP packet", SPINDUMP_IP_V(&ip));
                                                                               ^
/Users/lars/Documents/Code/spindump/src/spindump_protocols.h:63:32: note: expanded from macro 'SPINDUMP_IP_V'
#define SPINDUMP_IP_V(ip)     (((ip)->ip_vhl) >> 4)
                               ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:560:22: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
  uint8_t ecnFlags = SPINDUMP_IP_ECN(&ip);
                     ^
/Users/lars/Documents/Code/spindump/src/spindump_protocols.h:64:32: note: expanded from macro 'SPINDUMP_IP_ECN'
#define SPINDUMP_IP_ECN(ip)   (((ip)->ip_tos) & 0x3)
                               ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:567:8: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
  if ((off & SPINDUMP_IP_OFFMASK) != 0) {
       ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:569:57: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    spindump_debugf("ignored a fragment at offset %u", (off & SPINDUMP_IP_OFFMASK));
                                                        ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:631:23: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
  uint8_t ipVersion = SPINDUMP_IP6_V(&ip6);
                      ^
/Users/lars/Documents/Code/spindump/src/spindump_protocols.h:107:30: note: expanded from macro 'SPINDUMP_IP6_V'
#define SPINDUMP_IP6_V(ip)              (((ip)->ip6_vtc) >> 4)
                                         ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:634:80: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    spindump_warnf("IP versions inconsistent in Ethernet frame and IP packet", SPINDUMP_IP6_V(&ip6));
                                                                               ^
/Users/lars/Documents/Code/spindump/src/spindump_protocols.h:107:30: note: expanded from macro 'SPINDUMP_IP6_V'
#define SPINDUMP_IP6_V(ip)              (((ip)->ip6_vtc) >> 4)
                                         ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:661:22: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
  uint8_t ecnFlags = SPINDUMP_IP6_ECN(&ip6);
                     ^
/Users/lars/Documents/Code/spindump/src/spindump_protocols.h:108:42: note: expanded from macro 'SPINDUMP_IP6_ECN'
#define SPINDUMP_IP6_ECN(ip)            ((((ip)->ip6_tcfl >> 4)) & 0x3)
                                         ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:683:9: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    if (spindump_ip6_fh_fragoff(off) != 0) {
        ^
/Users/lars/Documents/Code/spindump/src/spindump_protocols.h:140:45: note: expanded from macro 'spindump_ip6_fh_fragoff'
# define spindump_ip6_fh_fragoff(field)    ((field)>>2)
                                            ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:685:59: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
      spindump_debugf("ignored a fragment at offset %u", (off & SPINDUMP_IP_OFFMASK));
                                                          ^
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:688:7: warning: do not use 'else' after 'return' [readability-else-after-return]
    } else {
      ^~~~~~
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:757:22: warning: statement should be inside braces [readability-braces-around-statements]
                        if (fromResponder)
                                          ^
                                           {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:759:8: warning: statement should be inside braces [hicpp-braces-around-statements]
                        else
                            ^
                             {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:764:22: warning: statement should be inside braces [readability-braces-around-statements]
                        if (fromResponder)
                                          ^
                                           {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:766:8: warning: statement should be inside braces [hicpp-braces-around-statements]
                        else
                            ^
                             {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:772:22: warning: statement should be inside braces [readability-braces-around-statements]
                        if (fromResponder)
                                          ^
                                           {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:774:8: warning: statement should be inside braces [readability-braces-around-statements]
                        else
                            ^
                             {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1101:18: warning: statement should be inside braces [google-readability-braces-around-statements]
  if (event == 0) return("none");
                 ^
                  {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1102:3: warning: do not use 'else' after 'return' [readability-else-after-return]
  else if (event == spindump_analyze_event_newconnection) return("newconnection");
  ^~~~~~~~
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1102:58: warning: statement should be inside braces [google-readability-braces-around-statements]
  else if (event == spindump_analyze_event_newconnection) return("newconnection");
                                                         ^
                                                          {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1103:61: warning: statement should be inside braces [hicpp-braces-around-statements]
  else if (event == spindump_analyze_event_connectiondelete) return("connectiondelete");
                                                            ^
                                                             {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1104:66: warning: statement should be inside braces [google-readability-braces-around-statements]
  else if (event == spindump_analyze_event_newleftrttmeasurement) return("newleftrttmeasurement");
                                                                 ^
                                                                  {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1105:67: warning: statement should be inside braces [hicpp-braces-around-statements]
  else if (event == spindump_analyze_event_newrightrttmeasurement) return("newrightrttmeasurement");
                                                                  ^
                                                                   {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1106:62: warning: statement should be inside braces [google-readability-braces-around-statements]
  else if (event == spindump_analyze_event_initiatorspinflip) return("initiatorspinflip");
                                                             ^
                                                              {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1107:62: warning: statement should be inside braces [hicpp-braces-around-statements]
  else if (event == spindump_analyze_event_responderspinflip) return("responderspinflip");
                                                             ^
                                                              {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1108:63: warning: statement should be inside braces [google-readability-braces-around-statements]
  else if (event == spindump_analyze_event_initiatorspinvalue) return("initiatorspinvalue");
                                                              ^
                                                               {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1109:63: warning: statement should be inside braces [google-readability-braces-around-statements]
  else if (event == spindump_analyze_event_responderspinvalue) return("responderspinvalue");
                                                              ^
                                                               {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1110:54: warning: statement should be inside braces [google-readability-braces-around-statements]
  else if (event == spindump_analyze_event_newpacket) return("newpacket");
                                                     ^
                                                      {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1111:64: warning: statement should be inside braces [hicpp-braces-around-statements]
  else if (event == spindump_analyze_event_firstresponsepacket) return("firstresponsepacket");
                                                               ^
                                                                {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1112:56: warning: statement should be inside braces [google-readability-braces-around-statements]
  else if (event == spindump_analyze_event_statechange) return("statechange");
                                                       ^
                                                        {
/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:1113:7: warning: statement should be inside braces [google-readability-braces-around-statements]
  else return("multiple");
      ^
       {
(information) Couldn't find path given by -I '/Users/lars/Documents/Code/spindump/src/src/'
Warning: cppcheck reported diagnostics:
[/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:254]: (style) The function 'spindump_analyze_getstats' is never used.
[/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:90]: (style) The function 'spindump_analyze_initialize' is never used.
[/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:327]: (style) The function 'spindump_analyze_process' is never used.
[/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:175]: (style) The function 'spindump_analyze_registerhandler' is never used.
[/Users/lars/Documents/Code/spindump/src/spindump_analyze.c:145]: (style) The function 'spindump_analyze_uninitialize' is never used.
(information) Cppcheck cannot find all the include files (use --check-config for details)
clang: error: argument unused during compilation: '-fsanitize-memory-track-origins' [-Werror,-Wunused-command-line-argument]
clang: error: argument unused during compilation: '-fsanitize-memory-use-after-dtor' [-Werror,-Wunused-command-line-argument]
clang: error: argument unused during compilation: '-fsanitize-address-use-after-scope' [-Werror,-Wunused-command-line-argument]
make[2]: *** [src/CMakeFiles/libspindump.dir/build.make:63: src/CMakeFiles/libspindump.dir/spindump_analyze.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:169: src/CMakeFiles/libspindump.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

**Comment by jariarkko** [comment]
Hmm. Clang gives more errors than gcc to begin with, seems like 1/ I should run also with clang (some further cleanup needed, clearly) 2/ we need to add -Wno-covered-switch-default because I decided that warning interferes too much with defensive programming that I prefer to use 3/ but even so, maybe there's some remaining clang or other version issue, because on my compilation I get the option errors even without any other errors. Maybe Clang version, OSX version, mac port vs. homebrew... many potential reasons for this. But we'd ideally have the cmake system work regardless of what specific versions there are... I'm working on this still, stay tuned :-)

**Comment by larseggert** [comment]
Your Xcode is older than mine, FWIW

**Comment by larseggert** [comment]
What Xcode calls gcc is really clang

**Comment by jariarkko** [comment]
The latest version of Spindumop now has cmake support (using code from this PR, although the PR itself was not taken forward). Thanks Lars!!!

**Pull request closed without merging**